clientapi/clientapi_test.go

```diff
@@ -0,0 +1,376 @@
+	"github.com/matrix-org/dendrite/internal/caching"
+	"github.com/matrix-org/dendrite/internal/httputil"
+	"github.com/matrix-org/dendrite/internal/sqlutil"
+	"github.com/matrix-org/dendrite/roomserver"
+	"github.com/matrix-org/dendrite/setup/jetstream"
+	"github.com/matrix-org/dendrite/test"
+	"github.com/matrix-org/dendrite/test/testrig"
+	"github.com/matrix-org/dendrite/userapi"
+	uapi "github.com/matrix-org/dendrite/userapi/api"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+	"github.com/tidwall/gjson"
+)
+
+type userDevice struct {
+	accessToken string
+	deviceID    string
+	password    string
+}
+
+func TestGetPutDevices(t *testing.T) {
+	alice := test.NewUser(t)
+	bob := test.NewUser(t)
+
+	test.WithAllDatabases(t, func(t *testing.T, dbType test.DBType) {
+		testCases := []struct {
+			name         string
+			requestUser  *test.User
+			deviceUser   *test.User
+			request      *http.Request
+			withDeviceID bool
+			wantOK       bool
+			validateFunc func(t *testing.T, device userDevice, routers httputil.Routers)
+		}{
+			{
+				name:        "can get all devices",
+				requestUser: alice,
+				request:     httptest.NewRequest(http.MethodGet, "/_matrix/client/v3/devices", strings.NewReader("")),
+				wantOK:      true,
+			},
+			{
+				name:         "can get specific own device",
+				requestUser:  alice,
+				deviceUser:   alice,
+				request:      httptest.NewRequest(http.MethodGet, "/_matrix/client/v3/devices/", strings.NewReader("")),
+				withDeviceID: true,
+				wantOK:       true,
+			},
+			{
+				name:         "can not get device for different user",
+				requestUser:  alice,
+				deviceUser:   bob,
+				request:      httptest.NewRequest(http.MethodGet, "/_matrix/client/v3/devices/", strings.NewReader("")),
+				withDeviceID: true,
+			},
+			{
+				name:         "can update own device",
+				requestUser:  alice,
+				deviceUser:   alice,
+				request:      httptest.NewRequest(http.MethodPut, "/_matrix/client/v3/devices/", strings.NewReader(`{"display_name":"my new displayname"}`)),
+				withDeviceID: true,
+				wantOK:       true,
+				validateFunc: func(t *testing.T, device userDevice, routers httputil.Routers) {
+					req := httptest.NewRequest(http.MethodGet, "/_matrix/client/v3/devices/"+device.deviceID, strings.NewReader(""))
+					req.Header.Set("Authorization", "Bearer "+device.accessToken)
+					rec := httptest.NewRecorder()
+					routers.Client.ServeHTTP(rec, req)
+					if rec.Code != http.StatusOK {
+						t.Fatalf("expected HTTP 200, got %d: %s", rec.Code, rec.Body.String())
+					}
+					gotDisplayName := gjson.GetBytes(rec.Body.Bytes(), "display_name").Str
+					if gotDisplayName != "my new displayname" {
+						t.Fatalf("expected displayname '%s', got '%s'", "my new displayname", gotDisplayName)
+					}
+				},
+			},
+			{
+				// this should return "device does not exist"
+				name:         "can not update device for different user",
```

Comment:
AKA a 404? Maybe swap `wantOK` with `wantStatusCode: 200` and then you can assert the HTTP status code for errors?